### ISSUE TITLE
[$1250] Webkit build for OS X

### ISSUE THREAD

**Issue by jmgasper** [topcoderinc/pxCore#316]
@jiangliwu - We have a webkit build linked in and working on Linux:

* https://github.com/krunt/pxCore/tree/_tc_video

Ubuntu build instructions: https://github.com/krunt/pxCore/blob/_tc_video/MSE_README_WEBKIT.md

What I'd like to do is get it working on MacOS as well, with Webkit built in, running the two demos listed, if possible.  

This would require changes to the README as well to ensure that all proper dependencies are documented properly.

**Comment by jmgasper** [comment]
Contest https://www.topcoder.com/challenges/30099984 has been created for this ticket.<br/><br/>```This is an automated message for ghostar via Topcoder X```

**Comment by jmgasper** [comment]
Contest https://www.topcoder.com/challenges/30099984 has been updated - it has been assigned to jiangliwu.<br/><br/>```This is an automated message for ghostar via Topcoder X```

**Comment by jiangliwu** [comment]
@jmgasper i follow linux build, but in MSEMediaSource.cpp `#include "WebCore/config.h"`, i see there is no `cmakeconfig.h` in whole webkit folder, do i missing something?

**Comment by jmgasper** [comment]
@jiangliwu - Did you run the updated build.sh in the externals folder?  Have a look at the bottom of that file - it should be pulling webkit, applying a patch, and then building it.

**Comment by jiangliwu** [comment]
@jmgasper yes, i am already successful build webkit, this errors are from building spark.

**Comment by jmgasper** [comment]
Strange.  Just to verify:

1.  You're on ubuntu 18.04
2.  The build.sh is completing successfully
3.  You're on the _tc_video branch

Can you attach the full build output please?

**Comment by jiangliwu** [comment]
yes, ubuntu is worked. i am working that on mac,  webkit lib build succeed, but spark build failed, maybe some different with ubuntu when build webkit, let me dig out the reason.

**Comment by jmgasper** [comment]
@jiangliwu - Any update here?  Anything I can help with?

**Comment by jiangliwu** [comment]
@jmgasper no more progress, let us release this

**Comment by jmgasper** [comment]
@krunt - This one's lower priority, but something to work on, if you want.

**Comment by jmgasper** [comment]
Contest https://www.topcoder.com/challenges/30099984 has been updated - it has been assigned to kruntuid.<br/><br/>```This is an automated message for ghostar via Topcoder X```

**Comment by jmgasper** [comment]
Contest https://www.topcoder.com/challenges/30099984 has been updated - the new changes has been updated for this ticket.<br/><br/>```This is an automated message for ghostar via Topcoder X```

**Comment by jmgasper** [comment]
@krunt - Got it to build and run using XCode 9 instead of 10.

Build for testing is here:

https://www.dropbox.com/s/64niriycgvk0963/pxScene_Webkit.zip?dl=1

When I try to load an MPD from the mse test JS app, I get this crash:

```
[SPARK LOG] error_data:[object Object]
[1661][ManifestUpdater] Manifest has been refreshed at Wed Sep 04 2019 08:08:56 GMT+1000 (AEST)[1567548536.01]

(<unknown>:23162): GLib-GObject-CRITICAL **: 08:08:56.083: g_object_set: assertion 'G_IS_OBJECT (object)' failed

** (<unknown>:23162): CRITICAL **: 08:08:56.086: gst_gl_display_x11_new_with_display: assertion 'display != NULL' failed
ASSERTION FAILED: m_glDisplay
../../Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp(454) : bool WebCore::MediaPlayerPrivateGStreamerBase::ensureGstGLContext()
1   0x1220f98b9 WTFCrash
2   0x10127a7db WTFCrashWithInfo(int, char const*, char const*, int)
3   0x10172333a WebCore::MediaPlayerPrivateGStreamerBase::ensureGstGLContext()
4   0x101722b90 WebCore::MediaPlayerPrivateGStreamerBase::requestGLContext(char const*)
5   0x101722355 WebCore::MediaPlayerPrivateGStreamerBase::handleSyncMessage(_GstMessage*)
6   0x10170687e WebCore::MediaPlayerPrivateGStreamer::handleSyncMessage(_GstMessage*)
7   0x101728ef4 WebCore::MediaPlayerPrivateGStreamerBase::setPipeline(_GstElement*)::$_0::operator()(_GstBus*, _GstMessage*, void*) const
8   0x101728eb5 WebCore::MediaPlayerPrivateGStreamerBase::setPipeline(_GstElement*)::$_0::__invoke(_GstBus*, _GstMessage*, void*)
9   0x119212e95 gst_bus_post
10  0x11922bae5 gst_element_post_message_default
11  0x1192066a2 gst_bin_post_message
12  0x1192289ac gst_element_post_message
13  0x12c2767a9 activate_sink_bus_handler
14  0x119212e95 gst_bus_post
15  0x11922bae5 gst_element_post_message_default
16  0x1192066a2 gst_bin_post_message
17  0x1192289ac gst_element_post_message
18  0x119208e5a gst_bin_handle_message_func
19  0x11920bd33 bin_bus_handler
20  0x119212e95 gst_bus_post
21  0x11922bae5 gst_element_post_message_default
22  0x1192066a2 gst_bin_post_message
23  0x1192289ac gst_element_post_message
24  0x119208e5a gst_bin_handle_message_func
25  0x11920bd33 bin_bus_handler
26  0x119212e95 gst_bus_post
27  0x11922bae5 gst_element_post_message_default
28  0x1192289ac gst_element_post_message
29  0x1190ab73a _gst_context_query
30  0x1190aa7c0 gst_gl_ensure_element_data
31  0x12c33073a gst_glimage_sink_change_state
Segmentation fault: 11
```

**Comment by krunt** [comment]
seems like no x11 display found
I fixed that on osx mojave 
running
sudo port -v install xorg-server

**Comment by krunt** [comment]
after that I've fixed crash in my repo
https://github.com/krunt/pxCore/commit/00ae713dacfe923cd8ff669ca47c3259b6fcb1fb
you can merge it and test afterwards
thanks

**Comment by jmgasper** [comment]
@krunt - Do we need X11?  That's not ideal.

**Comment by jmgasper** [comment]
@krunt - A couple things here:

1.  I installed X11 server via ports, (along with `xeyes`, for testing).  I have run X11 and xeyes and they work as expected.

2.  We really shouldn't need a dependency on X11.  That's not going to work for the client.

3.  I rebuilt the app and I'm now seeing a different error:

```
[61][MediaPlayer] [dash.js 3.0.0] MediaPlayer has been initialized
[1571][DashParser] Parsing complete: ( xml2json: 0.00ms, objectiron: 0.00ms, total: 0.00s)
[1589][StreamController] MediaSource attached to element.  Waiting on open...
[1589][ManifestUpdater] Manifest has been refreshed at Thu Sep 05 2019 08:52:33 GMT+1000 (AEST)[1567637553.446]

(<unknown>:1326): GLib-GObject-CRITICAL **: 08:52:33.493: g_object_set: assertion 'G_IS_OBJECT (object)' failed
dbus[1326]: Dynamic session lookup supported but failed: launchd did not provide a socket path, verify that org.freedesktop.dbus-session.plist is loaded!

(<unknown>:1326): GLib-CRITICAL **: 08:52:33.535: g_hash_table_insert_internal: assertion 'hash_table != NULL' failed

(<unknown>:1326): GLib-CRITICAL **: 08:52:33.535: g_hash_table_insert_internal: assertion 'hash_table != NULL' failed

(<unknown>:1326): GLib-CRITICAL **: 08:52:33.535: g_hash_table_insert_internal: assertion 'hash_table != NULL' failed

(<unknown>:1326): GLib-CRITICAL **: 08:52:33.535: g_hash_table_insert_internal: assertion 'hash_table != NULL' failed

(<unknown>:1326): GLib-GObject-WARNING **: 08:52:33.535: invalid cast from 'image/jpeg, parsed=(boolean)true' to 'GstGLDisplay'

(<unknown>:1326): Gdk-ERROR **: 08:52:33.539: The program '<unknown>' received an X Window System error.
This probably reflects a bug in the program.
The error was 'BadMatch (invalid parameter attributes)'.
  (Details: serial 151 error_code 8 request_code 148 (GLX) minor_code 27)
  (Note to programmers: normally, X errors are reported asynchronously;
   that is, you will receive the error a while after causing it.
   To debug your program, run it with the GDK_SYNCHRONIZE environment
   variable to change this behavior. You can then get a meaningful
   backtrace from your debugger if you break on the gdk_x_error() function.)
Trace/BPT trap: 5
```

**Comment by krunt** [comment]
I am currently building mac port of webkit on mojave (not gtk+ port), 
so maybe it will not require X11 (need check for this)

**Comment by jmgasper** [comment]
@krunt - Any luck?

**Comment by krunt** [comment]
seems like bug in opengl implementation on mojave (I have no luck fixing BadMatch),
the error in GLXCreatePbuffer function. I am currently building on highsierra but 
there are runtime gstreamer strange errors.

webkit configured with eight GLX or EGL, GLX is tightly connected with x11,
maybe we can try to build gtk+ webkit with EGL support and remove x11 as a dependency.

**Comment by jmgasper** [comment]
@krunt - Ok, a couple things:

1.  We did some initial work to integrate GStreamer here:  https://github.com/topcoderinc/pxCore/tree/_tc_video. Not sure if that woul dhelp with the gstreamer errors

2.  What GStreamer errors are you seeing?

**Comment by krunt** [comment]
1. I've reused that code for gl rendering (it does not help here).
2. The errors are inside webkit (MediaPlayerPrivateGStreamer.cpp):
the reason of crash:

GRefPtr<GstElement> m_pipeline = gst_element_factory_make("playbin", "media_player");
...
g_object_set(m_pipeline.get(), "audioSink", createAudioSink(), nullptr);
...
inside setSyncOnClock(audioSink(), sync);
-> g_object_get(m_pipeline.get(), "audio-sink", &sink, nullptr);
    sink is undefined here

The problem is can't set properties on m_pipeline (via g_object_set)
Strange warnings:
GLib-GObject-CRITICAL **: 09:51:19.067: g_param_spec_pool_lookup: assertion 'pool != NULL' failed
09:51:19.067: g_object_get_is_valid_property: object class '(null)' has no property named 'audio-sink'

I am installing gstreamer via macports
sudo port install webkit-gtk

**Comment by krunt** [comment]
it seems I have broken gstreamer installation

**Comment by krunt** [comment]
Also I've found issues on highsierra and mojave that multible libglib-2.0.0.dylib 
linked into application, 
one directly into Spark executable
another into libjavascriptcoregtk-4.0.18.dylib
this may be the cause of issues of gstreamer above ([here is mentioned](https://www.bountysource.com/issues/51445967-tutorial-3-4-on-macos-g_param_spec_pool_lookup-assertion-pool-null-failed))

on linux libglib is linked into Spark executable only,

I will try to investigate this

**Comment by krunt** [comment]
I finally got it all working!
Will commit in some time

**Comment by krunt** [comment]
I've committed fixes.
I tested on highsierra and it should work on mojave too.
1) To fix 2 glib libs linked in, you should not install gstreamer
through brew, only through port (I've updated build.sh to install through port)
Also after building 3rdparty before building Spark
ensure that  /usr/local/opt folder does not exist (remove or move it otherwise)
2) To fix BadMatch glx (its broken gl implementation I guess) I worked around by patching webkit
to use pixmap instead of pbuffer).

**Comment by jmgasper** [comment]
@krunt - Great news, thanks.  I'll mess around with this today / tomorrow.

**Comment by jmgasper** [comment]
Contest https://www.topcoder.com/challenges/30099984 has been updated - the new changes has been updated for this ticket.<br/><br/>```This is an automated message for ghostar via Topcoder X```

**Comment by jmgasper** [comment]
@krunt - Got the updated build working, after a lot of hairpulling with the dependencies.  Now seeing a crash, after I select a video to play:

[SparkLog.txt](https://github.com/topcoderinc/pxCore/files/3598854/SparkLog.txt)

**Comment by jmgasper** [comment]
Build I'm testing:  https://www.dropbox.com/s/n2ozagayv0nu7u5/Archive.zip?dl=1

**Comment by krunt** [comment]
In your build two libglib are linked in. 
Have you done this?
`Also after building 3rdparty before building Spark
ensure that /usr/local/opt folder does not exist (remove or move it otherwise)`
Anyway to fix your current build:
mv /usr/local/opt /usr/local/opt.1
mkdir /usr/local/opt
sudo cp -r /usr/local/opt.1/openssl /usr/local/opt
sudo cp -r /usr/local/opt.1/libffi /usr/local/opt
rebuild spark only
ensure you have one libglib and libgio linked in
otool -L Spark | grep libglib # here must be one line
otool -L Spark | grep libgio 

need to find a way to ensure this automatically

**Comment by jmgasper** [comment]
Thanks - I had done that move but moved it *back* due to other issues.

Now I'm seeing new errors - going to mess with it again tomorrow.

**Comment by jmgasper** [comment]
Contest https://www.topcoder.com/challenges/30099984 has been updated - the new changes has been updated for this ticket.<br/><br/>```This is an automated message for ghostar via Topcoder X```

**Comment by jmgasper** [comment]
Payment task has been updated: https://software.topcoder.com/review/actions/ViewProjectDetails?pid=30099984<br/><br/>```This is an automated message for ghostar via Topcoder X```
